docs/llama_multi_lora_tutorial.md

```diff
@@ -0,0 +1,201 @@
+    │   └── model.json
+    └── config.pbtxt
+```
+
+Now, you have finish the basic deployment, and the file path should look like this:
+
+```
+vllm_workspace
+├── weights
+│   ├── backbone
+│   │	└── llama-7b-hf
+│   └── loras
+│       ├── alpaca-lora-7b
+│       └── bactrian-x-llama-lora-7b
+│
+└── model_repository
+    └── vllm_model
+        ├── 1
+        │   └── model.json
+        └── config.pbtxt
+```
+
+
+
+__2.2 complete the `model.json` according to your needs__
+
+The content of `model.json` can be:
+
+```json
+{
+    "model":"/vllm_workspace/weights/backbone/llama-7b-hf",
+    "disable_log_requests": "true",
+    "gpu_memory_utilization": 0.8,
+    "tensor_parallel_size": 2,
+    "block_size": 16,
+    "enable_lora": "true",
+    "max_lora_rank": 16
+}
+```
+
++ `model`: The path to your model repository
++ `disable_log_requests`: To show logs when launch vllm or not.
++ `gpu_memory_utilization`: The gpu memory allocated for the model weights and vllm *PagedAttention* kv cache manager.
++ `tensor_parallel_size`: The vllm now support the tensor paralism, so you can decide how many gpus you want to use for serving.
++ `block_size`: vLLM kv cache block size.
++ `enable_lora`: If you want to support vllm multi-lora, this should be configured and set `true`.
++ `max_lora_rank`: The maximum of LoRA rank of your lora adapter.
+
+For more information, you may read the source code of [EngineArgs](https://github.com/Yard1/vllm/blob/multi_lora/vllm/engine/arg_utils.py#L11) here.
+
+
+
+__2.3 add `multi_lora.json` to figure out the local lora path__
+
+Now (2023.11.29) the [vLLM multi-lora PR](https://github.com/vllm-project/vllm/pull/1804) just supported the inference of **local lora weights applying**, which means that the vllm cannot pull any lora adapter from huggingface. So triton should know where the local loras weights are.
+
+Create a `multi_lora.json` file under `model_repository/vllm_model/1/` path:
+
+```bash
+cd model_repository/vllm_model/1
+touch multi_lora.json
+```
+
+A `multi_lora.json` should look like this:
+
+```json
+{
+    "alpaca": "/vllm_workspace/weights/loras/alpaca-lora-7b",
+    "bactrian": "/vllm_workspace/weights/loras/bactrian-x-llama-7b-lora"
+}
+```
+
+The **key** should be the supported lora name, and the **value** should be the specific path in your machine.
+
+
+
+## Step 3: Start a docker container for triton-vllm serving
+
+**A docker container is strongly recommended for serving**, and this tutorial will only demonstrate how to launch triton in docker env.
+
+First, create a docker container using the NCG built for vllm serving:
+
+```bash
+# NOTICE: you must first cd to your vllm_workspace path.
+cd vllm_workspace
+
+sudo docker run --gpus all -it --net=host -p 8001:8001 --shm-size=12G \
+--ulimit memlock=-1 --ulimit stack=67108864 -v ${PWD}:/vllm_workspace \
+-w /vllm_workspace nvcr.io/nvidia/tritonserver:23.10-vllm-python-py3 \
+/bin/bash
+```
+
+And download the `model.py` script from github:
```

Comment:
In case `src/model.py` is updated with `lora`-related changes, next steps are redundant. As an option, we can keep them, but mention that "For pre-[target release version] containers, please replace provided in the container `/opt/tritonserver/backends/vllm/model.py` with the most up to date version.

Plus, we need to mention at what version of vllm `lora` support was added, since 23.10 container was build with 0.2.0 I believe, or later. 23.12 container comes with vllm version 0.2.3

Reply:
The current vllm version is 0.2.6, and multi-lora PR (by Yard1) is not merged yet. (multi-lora feature will be added in vllm 0.2.7 or later)

Since 23.12 container integrates vllm 0.2.3, the multi-lora feature will not be added temporarily. (Maybe `24.yy` container will support vllm 0.2.7+ version)

Reply:
I have understood what you mean, docs updated.